client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-sections-map-by-category.ts

```diff
@@ -0,0 +1,21 @@
+import { useMemo } from 'react';
+import type { Pattern, Category } from '../types';
+
+const useSectionsMapByCategory = ( patterns: Pattern[], categories: Category[] ) => {
+	return useMemo(
+		() =>
+			categories.reduce( ( categoriesMap: { [ key: string ]: Pattern[] }, category ) => {
```

Comment:
If we traverse `patterns` inside `categories.reduce`, we will go through the pattern list sooo many times even if this calculation might be memo forever.

Instead, we can create the `categoriesMap` first, and then traverse `patterns` to push it to its categories. For example,

```js
const categoriesMap = categories.reduce( ( acc, cur ) => ( {
  ...acc,
  [ cur.name ]: [],
} );

patterns.forEach( pattern => {
  patterns.categories.forEach( category => {
    if ( categoriesMap[ category ] ) {
      categoriesMap[ category ].push( pattern );
    }
  } );
} );
```